runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.metrics.micrometer.MicroMeterUtils;
+import org.corfudb.protocols.wireprotocol.StreamAddressRange;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.collections.StreamListener;
+import org.corfudb.runtime.exceptions.StreamingException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.SequencerView;
+import org.corfudb.runtime.view.stream.StreamAddressSpace;
+
+import javax.annotation.Nonnull;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+/**
+ * This class implements a scheduler that manages multiple streaming tasks {@link StreamingTask}. A streaming task
+ * is simply a listener attached to a corfu stream and expects to receive ordered change notifications for certain
+ * tags of interest. The StreamPollingScheduler manages the life cycle of streaming tasks, this constitutes four
+ * major states, polling a stream, reading deltas, notifying the listeners of the changes and propagating errors.
+ * Each stream is managed according to the diagram below:
+ *
+ *                    │ Register
+ *           ┌────────▼──────────┐
+ *           │                   │◄───┐
+ *    ┌─────►│     Runnable      │    │
+ *    │      └─────────┬─────────┘    │
+ *    │                │            No│Deltas
+ *    │                │ Poll         │
+ *    │                ▼              │
+ *    │      ┌───────────────────┐    │
+ * Synced    │    Scheduling     ├────┘
+ *    │      └─────────┬─────────┘
+ *    │                │
+ *    │                │Dispatch
+ *    │                ▼
+ *    │       ┌───────────────────┐
+ *    │       │                   │ Produce
+ *    └───────┤      Syncing      │◄────┐
+ *            └────────┬──────────┴─────┘
+ *                     │ Exception
+ *                     ▼
+ *            ┌───────────────────┐
+ *            │      Error        │
+ *            └───────────────────┘
+ *
+ * On every tick the scheduler will poll/dispatch tasks only when needed, for example, when a listener is not
+ * consuming the deltas fast enough the scheduler can decide to not poll or schedule extra sync tasks until
+ * it can, this prevents superfluous polling and overwhelming the listener. Furthermore, syncing and listener
+ * notification is executed on a shared thread pool in round robin fashion, therefore if a stream produces
+ * more deltas relative to other streams that particular stream will have more CPU time than other tasks.
+ */
+
+
+@Slf4j
+public class StreamPollingScheduler {
```

Comment:
BTW, did you consider using quartz-scheduler instead of self written code?

Reply:
Or [RxJava](https://www.baeldung.com/rxjava-schedulers) even better alternative

Reply:
I haven't considered using quartz, but I think it would be an overkill. 

Which part of this patch do you think would be simplified/cleaned/optimized if we opt to use quartz?

Reply:
Since the `AddressSpaceView::read` and `SequencerView:getStreamAddressSpace` are blocking APIs I don't see any performance benefit of using RxJava. Since this will be loaded on the client side, I don't want to add dependencies unless absolutely necessary.

If you have a specific idea on how to improve this schedulers implementation using RxJava please feel free to articulate it here and I will investigate it.